components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/AbstractUserInfoResponseBuilder.java

```diff
@@ -63,12 +63,33 @@ public String getResponseString(OAuth2TokenValidationResponseDTO tokenResponse)
 
         String clientId = getClientId(OAuth2Util.getAccessTokenIdentifier(tokenResponse));
         String spTenantDomain = getServiceProviderTenantDomain(tokenResponse);
+        OAuthAppDO oAuthAppDO;
+        try {
+            oAuthAppDO = OAuth2Util.getAppInformationByClientId(clientId, spTenantDomain);
+        } catch (IdentityOAuth2Exception | InvalidOAuthClientException e) {
+            throw new UserInfoEndpointException("Error while getting subject claim for client_id: " + clientId +
+                    " of tenantDomain: " + spTenantDomain, e);
+        }
         // Retrieve user claims.
         Map<String, Object> userClaims = retrieveUserClaims(tokenResponse);
         Map<String, Object> filteredUserClaims = filterOIDCClaims(tokenResponse, clientId, spTenantDomain, userClaims);
 
         // Handle subject claim.
         String subjectClaim = getSubjectClaim(userClaims, clientId, spTenantDomain, tokenResponse);
+        String callbackUri;
+        try {
+            callbackUri = OIDCClaimUtil.getCallbackUrl(clientId, spTenantDomain);
+        } catch (IdentityOAuth2Exception | InvalidOAuthClientException e) {
+            throw new UserInfoEndpointException("Error while getting the registered callback URI for client_id: " +
+                    clientId, e);
+        }
+        try {
+            // Get subject identifier according to the configured subject type.
+            subjectClaim = OIDCClaimUtil.getSubjectClaim(subjectClaim, oAuthAppDO);
```

Comment:
Let bring getting the subjectClaim to a private method and add the OAuthAppDO and callback uri resolution logic also into that private method as those two are not needed any where else